File: ios/Packages/ChainServices/ChainService/ServiceStatusService.swift
// Copyright (c). Gem Wallet. All rights reserved.

import protocol Gemstone.AlienProvider
import struct Gemstone.GemServiceEndpoint
import class Gemstone.GemServiceStatus

public struct ServiceStatusService: ServiceStatusServiceable, Sendable {
    private let client: GemServiceStatus

    public init(provider: any AlienProvider) {
        self.client = GemServiceStatus(provider: provider)
    }

    public var endpoints: [GemServiceEndpoint] {
        client.getEndpoints()
    }

    public func endpointLatency(url: String) async throws -> UInt64 {
        try await client.getEndpointLatency(url: url)
    }
}

File: android/blockchain/src/main/kotlin/com/gemwallet/android/blockchain/services/ServiceStatusService.kt
package com.gemwallet.android.blockchain.services

import kotlinx.coroutines.CancellationException
import uniffi.gemstone.GemServiceEndpoint
import uniffi.gemstone.GemServiceStatus

class ServiceStatusService(
    private val client: GemServiceStatus,
) {

    fun getEndpoints(): List<GemServiceEndpoint> = client.getEndpoints()

    suspend fun getEndpointLatency(url: String): ULong? {
        return try {
            client.getEndpointLatency(url)
        } catch (error: CancellationException) {
            throw error
        } catch (_: Throwable) {
            null
        }
    }
}
